charmcraft/services/package.py

```diff
@@ -239,10 +239,17 @@ def write_metadata(self, path: pathlib.Path) -> None:
 
         project_dict = self._project.marshal()
 
-        if (path / "metadata.yaml").exists():
+        # If there is a reactive part, defer to it for the existence of metadata.yaml.
+        plugins = {part.get("plugin") or name for name, part in self._project.parts.items()}
+        if (
+            "reactive" in plugins
+            and (
+                self._services.lifecycle.project_info.dirs.stage_dir / const.METADATA_FILENAME
```

Comment:
consider putting this `stage_dir` in a separate variable to make this `if` easier to read